UploadDaemon/SymbolAnalysis/LineCoverageSynthesizer.cs

```diff
@@ -15,10 +15,13 @@ public class LineCoverageSynthesizer : ILineCoverageSynthesizer
     {
         private static readonly Logger logger = LogManager.GetCurrentClassLogger();
 
+        private readonly IDictionary<string, IDictionary<GlobPatternList, SymbolCollection>> symbolCollectionsCache = new Dictionary<string, IDictionary<GlobPatternList, SymbolCollection>>();
```

Comment:
this data structure is very hard to comprehend. If I get this correctly, you want to use the symbol directory and the glob pattern list as a composite key to index the cached symbol collections, right?
For the sake of readability, I'd make this explicit: I think you can use tuples as Dictionary keys, which would a) make the intent clearer and b) simplify the code to access the dictionary.
Also, I'd move the cache into a separate class and use it from the SymbolCollection instead of from here. It seems to me the synthesizer shouldn't care whether it is using fresh or cached symbols. It just wants the symbols. So the Collection can handle the caching transparently
Making it its own class also allows us to add cache invalidation. Because one problem I see with this implementation is: if they redeploy their application and pdbs but don't kill the upload daemon, the deamon will incorrectly use the cached, outdated symbols against the new trace files. This will result in errors and incorrect coverage. We could e.g. in addition to the symbols cache some meta info about the pdbs (e.g. file names, last modified dates, file sizes?) and invalidate the cache if these don't match what we find on disk. Of course, we'd have to time how long gathering the meta data takes - would be bad if that's the bottleneck and not parsing the pdbs. Maybe some metadata can be read rather fast and some is slow (e.g. file size?)

Reply:
You got my intent right :)

I switched to using a `Tuple` and moved the caching into a dedicated `Resolver` class. Thanks for these suggestions.

I'm unsure about the invalidation. Could we somehow achieve invalidation by looking at the traces? The original issue was that reading the PDB files from a network share is slow. I fear if we introduce any reading from this source with every resolution, performance would regress here...

Reply:
I don't see how. The traces know nothing about PDBs. If they really have such a slow network share that listing the modified date of a few dozen files takes too long, I'd suggest not storing the PDBs on the network share ;-)
Could add a "don't invalidate pdb cache" option that is "use at your own risk". The default behaviour IMO **must** be that we properly invalidate the cache as otherwise we get very hard to debug problems with incorrect coverage. I don't want to go down that road ;-)

Reply:
Implemented cache invalidation based on the `LastWriteDate`. Struggled somewhat to write a test for this that runs both on my machine and on AppVeyor (because setting the `LastWriteDate` has some funny side effects, such as setting the date to `Now` if called multiple times consequently...), but it seems I succeeded :)